### PR TITLE
[build-utils] Add zero config detection for bun package manager

### PR DESCRIPTION
**Pull request by styfle**
> [!IMPORTANT]  
> This PR will only support Bun as a package manager at build time. 
> Bun will **not** work at runtime with Serverless Functions or Edge Functions at this time.

- Depends on https://github.com/vercel/api/pull/21869
- Fixes https://github.com/orgs/vercel/discussions/2021
- Closes https://github.com/vercel/vercel/pull/10244
- Related https://github.com/nodejs/corepack/issues/295
- Docs https://bun.sh/docs/install

### CODE REVIEW

**Review by trek — DISMISSED** [review]

**Comment by isaac-mcfadyen** [comment]
I saw that this was previously opened in https://github.com/vercel/vercel/pull/10244 and in fact was *already* closed with:
> We don't currently plan to support bun as a package manager or a runtime

Just curious where things changed here to make support for Bun a wanted feature?

**Comment by EndangeredMassa** [comment]
@isaac-mcfadyen yes, sorry for the confusion. At the time of your previous PR, we had not received many requests for `bun` support (even as just the package manager). As you can imagine, the release of `1.0.0` caused a lot of excitement in the growing community, which lead to [an increase in requests](https://github.com/orgs/vercel/discussions/2021).

File: packages/build-utils/src/fs/run-user-scripts.ts [hunk 1]
```diff
@@ -315,6 +323,11 @@ export async function scanParentDirs(
     cliType = 'npm';
     lockfilePath = npmLockPath;
     lockfileVersion = packageLockJson.lockfileVersion;
+  } else if (bunLockBin) {
+    cliType = 'bun';
+    lockfilePath = bunLockPath;
+    // TODO: read "bun-lockfile-format-v0"
```
Comment (Ethan-Arrowood) [inline packages/build-utils/src/fs/run-user-scripts.ts:329]: Lets not forget to make a linear ticket for this

**Comment by styfle** [comment]
> Just curious where things changed here to make support for Bun a wanted feature?

The [feature request](https://github.com/orgs/vercel/discussions/2021) had maybe 50 upvotes yesterday.

Today it has 750+ upvotes, likely because of [this tweet](https://twitter.com/jarredsumner/status/1701088829145518540).

File: packages/build-utils/src/fs/run-user-scripts.ts (hunk 1 above)
Comment (styfle) [reply packages/build-utils/src/fs/run-user-scripts.ts:329]: We can create a ticket when Bun 2.0 comes out or if the lockfile changes (hopfully that would be semver major but you never know).

I asked here https://github.com/orgs/vercel/discussions/2021#discussioncomment-6968013

**Comment by styfle** [comment]
Hmm, there appears to be a bug with Bun.

I filed an issue here: https://github.com/oven-sh/bun/issues/5007

**Review by TooTallNate — APPROVED** [review]

**Review by cb1kenobi — APPROVED** [review]

**Pull request merged**